File: agent/utils/openrouter-llm.ts
import { BaseLlm, LlmRequest, LlmResponse, GOOGLE_SEARCH, LLMRegistry } from '@google/adk';
import { logger } from './logger';

/**
 * Recursively redacts or truncates object keys and values (like 'content', 'text', 'arguments', 'args')
 * that might contain user data or heavy model payloads to keep log files safe and compact.
 *
 * @param value - The input value to redact/truncate.
 * @returns The sanitized value.
 */
function redactOrTruncate(value: any): any {
    if (typeof value === 'string') {
        return value.length > 100 ? value.substring(0, 100) + '... [truncated]' : value;
    }
    if (value && typeof value === 'object') {
        const result: any = Array.isArray(value) ? [] : {};
        for (const [key, val] of Object.entries(value)) {
            if (['content', 'text', 'arguments', 'args'].includes(key)) {
                result[key] =
                    typeof val === 'string'
                        ? val.length > 100
                            ? val.substring(0, 100) + '... [truncated]'
                            : val
                        : '[redacted/truncated object]';
            } else {
                result[key] = redactOrTruncate(val);
            }
        }
        return result;
    }
    return value;
}

// Override the processLlmRequest of GOOGLE_SEARCH to prevent it from throwing for non-Gemini models
const originalProcessLlmRequest = GOOGLE_SEARCH.processLlmRequest;
GOOGLE_SEARCH.processLlmRequest = async function (params) {
    const { llmRequest } = params;
    if (!llmRequest || !llmRequest.model) {
        return;
    }

    const openRouterModelRegex = /^(gemini-)?([a-zA-Z0-9\-]+\/[a-zA-Z0-9\-]+.*)$/;
    if (openRouterModelRegex.test(llmRequest.model)) {
        // For OpenRouter models, initialize config/tools and add placeholder to prevent crashes
        llmRequest.config = llmRequest.config || {};
        llmRequest.config.tools = llmRequest.config.tools || [];

        const hasSearch = llmRequest.config.tools.some(
            (t: any) => t.googleSearch || t.googleSearchRetrieval,
        );
        if (!hasSearch) {
            llmRequest.config.tools.push({
                googleSearch: {},
            });
        }
        return;
    }

    return originalProcessLlmRequest.call(this, params);
};

/**
 * OpenRouterLlm class implements a custom BaseLlm provider to integrate OpenRouter
 * compatible LLMs (like deepseek or mistral) into the Google ADK runner.
 */
export class OpenRouterLlm extends BaseLlm {
    // Matches model IDs that have a slash (e.g. google/gemini-2.5-flash) or openrouter/ prefixed ones, and allows gemini- prefix to bypass adk checks
    static supportedModels = [/^(gemini-)?([a-zA-Z0-9\-]+\/[a-zA-Z0-9\-]+.*)$/];

    /**
     * Creates an instance of OpenRouterLlm.
     *
     * @param params - Configuration parameters including the model string.
     */
    constructor(params: { model: string }) {
        super({ model: params.model });
    }

    /**
     * Sets up a websocket connection.
     * Note: WebSocket stream is not supported for OpenRouter, so this always throws.
     *
     * @param llmRequest - The parameters of the LLM request.
     * @throws Error indicating websocket is not supported.
     */
    async connect(llmRequest: LlmRequest): Promise<any> {
        throw new Error('WebSocket connection (connect) is not supported for OpenRouterLlm');
    }

    /**
     * Sends the chat request to OpenRouter API and yields incremental responses.
     *
     * @param llmRequest - The structured request payload parameters.
     * @param stream - Flag indicating whether to stream the LLM response.
     * @param abortSignal - Signal to abort the API request.
     * @returns An async generator yielding structured ADK LlmResponse chunks.
     */
    async *generateContentAsync(
        llmRequest: LlmRequest,
        stream: boolean = false,
        abortSignal?: AbortSignal,
    ): AsyncGenerator<LlmResponse, void> {
        const { url, headers, body } = this.buildRequestPayload(llmRequest, stream);

        const response = await fetch(url, {
            method: 'POST',
            headers,
            body: JSON.stringify(body),
            signal: abortSignal,
        });

        if (!response.ok) {
            const errorText = await response.text();
            throw new Error(`OpenRouter API error (${response.status}): ${errorText}`);
        }

        if (stream) {
            yield* this.handleStreamResponse(response);
        } else {
            yield* this.handleSingleResponse(response);
        }
    }

    /**
     * Maps the Google ADK Gemini-style contents/messages history structure
     * into OpenAI-style system, user, assistant, and tool messages list.
     *
     * @param contents - The list of contents to map.
     * @param systemInstruction - Optional system prompt instruction.
     * @returns A mapped list of OpenAI message objects.
     */
    private mapMessages(contents: LlmRequest['contents'], systemInstruction?: any): any[] {
        const messages: any[] = [];

        // 1. Add System Instruction
        if (systemInstruction) {
            messages.push({
                role: 'system',
                content: systemInstruction,
            });
        }

        // 2. Map Gemini contents to OpenAI-style messages
        for (const content of contents) {
            // Map role: 'model' -> 'assistant', 'user' -> 'user'
            const role = content.role === 'model' ? 'assistant' : 'user';

            let text = '';
            const toolCalls: any[] = [];

            for (const part of content.parts || []) {
                if (part.text) {
                    text += part.text;
                } else if (part.functionCall) {
                    // Map Gemini functionCall to OpenAI tool_calls
                    toolCalls.push({
                        id:
                            part.functionCall.id ||
                            `call_${Math.random().toString(36).substr(2, 9)}`,
                        type: 'function',
                        function: {
                            name: part.functionCall.name,
                            arguments: JSON.stringify(part.functionCall.args),
                        },
                    });
                } else if (part.functionResponse) {
                    // Map Gemini functionResponse to OpenAI tool role message
                    messages.push({
                        role: 'tool',
                        tool_call_id: part.functionResponse.id,
                        name: part.functionResponse.name,
                        content: JSON.stringify(part.functionResponse.response),
                    });
                }
            }

            // Add user/assistant message if it contains text or tool calls
            if (text || toolCalls.length > 0) {
                const msg: any = { role };
                if (text) {
                    msg.content = text;
                }
                if (toolCalls.length > 0) {
                    msg.tool_calls = toolCalls;
                }
                messages.push(msg);
            }
        }

        return messages;
    }

    /**
     * Maps Google ADK Gemini-style function declarations and search tools
     * to OpenAI-style tools/functions payload.
     *
     * @param tools - The ADK tools array.
     * @returns A mapped list of tool definitions.
     */
    private mapTools(tools?: any[]): any[] {
        const mappedTools: any[] = [];
        if (tools) {
            for (const t of tools) {
                const anyTool = t as any;
                if (anyTool.functionDeclarations) {
                    for (const decl of anyTool.functionDeclarations) {
                        mappedTools.push({
                            type: 'function',
                            function: {
                                name: decl.name,
                                description: decl.description || '',
                                parameters: decl.parameters || { type: 'object', properties: {} },
                            },
                        });
                    }
                } else if (anyTool.googleSearch || anyTool.googleSearchRetrieval) {
                    // Map Gemini googleSearch/googleSearchRetrieval to OpenRouter's web_search tool
                    mappedTools.push({
                        type: 'openrouter:web_search',
                    });
                }
            }
        }
        return mappedTools;
    }

    /**
     * Maps response MIME type configurations to OpenAI response formats (e.g., JSON mode).
     *
     * @param responseMimeType - Optional response mime type string.
     * @returns A response format object or undefined.
     */
    private mapResponseFormat(responseMimeType?: string): any {
        if (responseMimeType === 'application/json') {
            return { type: 'json_object' };
        }
        return undefined;
    }

    /**
     * Builds the final HTTP request URL, headers, and request body payload
     * needed for querying the OpenRouter completions endpoint.
     *
     * @param llmRequest - The incoming ADK request structure.
     * @param stream - Whether the request will be streamed.
     * @returns An object containing the target URL, headers, and raw request body.
     */
    private buildRequestPayload(
        llmRequest: LlmRequest,
        stream: boolean,
    ): { url: string; headers: Record<string, string>; body: any } {
        const messages = this.mapMessages(
            llmRequest.contents,
            llmRequest.config?.systemInstruction,
        );
        const tools = this.mapTools(llmRequest.config?.tools);
        const responseFormat = this.mapResponseFormat(llmRequest.config?.responseMimeType);

        const openrouterUrl = (process.env.OPENROUTER_URL || 'https://openrouter.ai').replace(
            /\/$/,
            '',
        );
        const apiKey = process.env.OPENROUTER_API_KEY;

        if (!apiKey) {
            throw new Error('OPENROUTER_API_KEY environment variable is not set');
        }

        const cleanModel = this.model.startsWith('gemini-') ? this.model.substring(7) : this.model;

        const requestBody: any = {
            model: cleanModel,
            messages,
            stream,
        };

        if (tools.length > 0) {
            requestBody.tools = tools;
        }

        if (responseFormat) {
            requestBody.response_format = responseFormat;
        }

        const headers: Record<string, string> = {
            'Content-Type': 'application/json',
            Authorization: `Bearer ${apiKey}`,
            'HTTP-Referer': 'https://github.com/aaochoa/study-buddy',
            'X-Title': 'Study Buddy AI',
        };

        return { url: `${openrouterUrl}/api/v1/chat/completions`, headers, body: requestBody };
    }

    /**
     * Handles SSE streaming responses from the OpenRouter API, accumulating
     * text deltas and tool call increments, yielding them in ADK format.
     *
     * @param response - The HTTP Response stream.
     * @returns An async generator yielding structured ADK LlmResponse chunks.
     */
    private async *handleStreamResponse(response: Response): AsyncGenerator<LlmResponse, void> {
        if (!response.body) {
            throw new Error('Response body is empty');
        }

        const reader = response.body.getReader();
        const decoder = new TextDecoder('utf-8');
        let buffer = '';
        let accumulatedText = '';

        const accumulatedToolCalls: Record<
            number,
            {
                id?: string;
                name?: string;
                arguments: string;
            }
        > = {};

        let finishReason: string | undefined = undefined;

        while (true) {
            const { done, value } = await reader.read();
            if (done) break;

            buffer += decoder.decode(value, { stream: true });
            const lines = buffer.split('\n');
            buffer = lines.pop() || '';

            for (const line of lines) {
                const trimmed = line.trim();
                if (!trimmed) continue;
                if (trimmed.startsWith('data: ')) {
                    const dataStr = trimmed.slice(6);
                    if (dataStr === '[DONE]') {
                        break;
                    }
                    try {
                        const parsed = JSON.parse(dataStr);
                        const choice = parsed.choices?.[0];
                        if (!choice) continue;

                        if (choice.finish_reason) {
                            finishReason =
                                choice.finish_reason === 'stop' ? 'STOP' : choice.finish_reason;
                        }

                        const delta = choice.delta;
                        if (!delta) continue;

                        const parts: any[] = [];

                        if (delta.content) {
                            accumulatedText += delta.content;
                            parts.push({ text: delta.content });
                        }

                        if (delta.tool_calls) {
                            for (const tc of delta.tool_calls) {
                                const index = tc.index;
                                if (accumulatedToolCalls[index] === undefined) {
                                    accumulatedToolCalls[index] = { arguments: '' };
                                }
                                const acc = accumulatedToolCalls[index];
                                if (tc.id) acc.id = tc.id;
                                if (tc.function?.name) acc.name = tc.function.name;
                                if (tc.function?.arguments) acc.arguments += tc.function.arguments;
                            }
                        }

                        if (parts.length > 0) {
                            yield {
                                content: {
                                    role: 'model',
                                    parts,
                                },
                                partial: true,
                            };
                        }
                    } catch (e) {
                        // Ignore partial JSON parsing errors
                    }
                }
            }
        }

        // Yield any accumulated tool calls once stream ends
        const finalParts: any[] = [];
        for (const index of Object.keys(accumulatedToolCalls)
            .map(Number)
            .sort((a, b) => a - b)) {
            const acc = accumulatedToolCalls[index];
            if (acc.name) {
                let args = {};
                try {
                    args = JSON.parse(acc.arguments);
                } catch (e) {
                    logger.error(
                        { arguments: redactOrTruncate(acc.arguments), err: e },
                        'Failed to parse accumulated tool call arguments',
                    );
                }
                finalParts.push({
                    functionCall: {
                        name: acc.name,
                        args,
                        id: acc.id,
                    },
                });
            }
        }

        // Always yield a final non-partial event containing the full accumulated text or tools
        yield {
            content: {
                role: 'model',
                parts: finalParts.length > 0 ? finalParts : [{ text: accumulatedText }],
            },
            finishReason: (finishReason || 'STOP') as any,
            partial: false,
        };
    }

    /**
     * Handles non-streaming responses from the OpenRouter API, mapping the
     * completions result and any function calls to the ADK LlmResponse format.
     *
     * @param response - The HTTP Response.
     * @returns An async generator yielding a single final ADK LlmResponse.
     */
    private async *handleSingleResponse(response: Response): AsyncGenerator<LlmResponse, void> {
        const data = await response.json();
        const choice = data.choices?.[0];
        const message = choice?.message;

        if (!message) {
            logger.error(
                { responseData: redactOrTruncate(data) },
                'OpenRouter full response on failure',
            );
            throw new Error('No message returned from OpenRouter');
        }

        const parts: any[] = [];

        if (message.content) {
            parts.push({ text: message.content });
        }

        if (message.tool_calls) {
            for (const tc of message.tool_calls) {
                if (tc.type === 'function') {
                    let args = {};
                    try {
                        args = JSON.parse(tc.function.arguments);
                    } catch (e) {
                        logger.error(
                            { arguments: redactOrTruncate(tc.function.arguments), err: e },
                            'Failed to parse function arguments',
                        );
                    }
                    parts.push({
                        functionCall: {
                            name: tc.function.name,
                            args,
                            id: tc.id,
                        },
                    });
                }
            }
        }

        yield {
            content: {
                role: 'model',
                parts,
            },
            finishReason: (choice.finish_reason === 'stop' ? 'STOP' : choice.finish_reason) as any,
            partial: false,
        };
    }
}

// Register the custom OpenRouter LLM provider once globally
LLMRegistry.register(OpenRouterLlm);

File: agent_python/utils/openrouter_llm.py
import os
import json
import httpx
from typing import AsyncGenerator, List, Dict, Any, Optional, ClassVar
from google.adk.models import BaseLlm, LlmRequest, LlmResponse, LLMRegistry
from google.genai import types
from utils.logger import logger

def redact_or_truncate(value: Any) -> Any:
    if isinstance(value, str):
        return value[:100] + "... [truncated]" if len(value) > 100 else value
    if isinstance(value, dict):
        res = {}
        for k, v in value.items():
            if k in ["content", "text", "arguments", "args"]:
                res[k] = (v[:100] + "... [truncated]") if isinstance(v, str) and len(v) > 100 else "[truncated]"
            else:
                res[k] = redact_or_truncate(v)
        return res
    if isinstance(value, list):
        return [redact_or_truncate(x) for x in value]
    return value

class OpenRouterLlm(BaseLlm):
    @classmethod
    def supported_models(cls) -> List[str]:
        return [r"^(gemini-)?([a-zA-Z0-9\-]+\/[a-zA-Z0-9\-]+.*)$"]

    def __init__(self, model: str, **kwargs):
        super().__init__(model=model, **kwargs)

    async def connect(self, llm_request: LlmRequest) -> Any:
        raise NotImplementedError("Live WebSocket connection is not supported for OpenRouterLlm")

    async def generate_content_async(
        self, llm_request: LlmRequest, stream: bool = False
    ) -> AsyncGenerator[LlmResponse, None]:
        url, headers, body = self._build_request_payload(llm_request, stream)

        async with httpx.AsyncClient(timeout=120.0) as client:
            if stream:
                async with client.stream("POST", url, headers=headers, json=body) as response:
                    if response.status_code != 200:
                        error_text = await response.aread()
                        raise RuntimeError(f"OpenRouter API error ({response.status_code}): {error_text.decode('utf-8')}")
                    async for item in self._handle_stream_response(response):
                        yield item
            else:
                response = await client.post(url, headers=headers, json=body)
                if response.status_code != 200:
                    raise RuntimeError(f"OpenRouter API error ({response.status_code}): {response.text}")
                for item in self._handle_single_response(response.json()):
                    yield item

    def _map_messages(self, contents: List[Any], system_instruction: Any = None) -> List[Dict[str, Any]]:
        messages = []

        # System instruction
        sys_text = ""
        if system_instruction:
            if isinstance(system_instruction, str):
                sys_text = system_instruction
            elif hasattr(system_instruction, "parts"):
                parts = system_instruction.parts or []
                sys_text = "".join([p.text for p in parts if hasattr(p, "text") and p.text])
            elif isinstance(system_instruction, dict):
                sys_text = system_instruction.get("text", "")

        if sys_text:
            messages.append({"role": "system", "content": sys_text})

        for content in contents or []:
            role = "assistant" if getattr(content, "role", None) == "model" else "user"
            text = ""
            tool_calls = []

            parts = getattr(content, "parts", []) or []
            for part in parts:
                part_text = getattr(part, "text", None)
                if part_text:
                    text += part_text
                
                # Check function call
                fn_call = getattr(part, "function_call", None)
                if fn_call:
                    call_id = getattr(fn_call, "id", None) or f"call_{os.urandom(4).hex()}"
                    args = getattr(fn_call, "args", {}) or {}
                    tool_calls.append({
                        "id": call_id,
                        "type": "function",
                        "function": {
                            "name": getattr(fn_call, "name", ""),
                            "arguments": json.dumps(args) if isinstance(args, dict) else str(args)
                        }
                    })

                # Check function response
                fn_resp = getattr(part, "function_response", None)
                if fn_resp:
                    resp_id = getattr(fn_resp, "id", "")
                    fn_name = getattr(fn_resp, "name", "")
                    resp_data = getattr(fn_resp, "response", {})
                    messages.append({
                        "role": "tool",
                        "tool_call_id": resp_id,
                        "name": fn_name,
                        "content": json.dumps(resp_data) if isinstance(resp_data, dict) else str(resp_data)
                    })

            if text or tool_calls:
                msg: Dict[str, Any] = {"role": role}
                if text:
                    msg["content"] = text
                if tool_calls:
                    msg["tool_calls"] = tool_calls
                messages.append(msg)

        return messages

    def _map_tools(self, tools: Any = None) -> List[Dict[str, Any]]:
        mapped_tools = []
        if not tools:
            return mapped_tools

        tool_list = tools if isinstance(tools, list) else [tools]
        for t in tool_list:
            fn_decls = getattr(t, "function_declarations", None)
            if fn_decls:
                for decl in fn_decls:
                    mapped_tools.append({
                        "type": "function",
                        "function": {
                            "name": getattr(decl, "name", ""),
                            "description": getattr(decl, "description", "") or "",
                            "parameters": getattr(decl, "parameters", {}) or {"type": "object", "properties": {}}
                        }
                    })
            elif getattr(t, "google_search", None) or getattr(t, "google_search_retrieval", None):
                mapped_tools.append({"type": "openrouter:web_search"})

        return mapped_tools

    def _build_request_payload(self, llm_request: LlmRequest, stream: bool) -> tuple[str, dict, dict]:
        sys_inst = None
        if hasattr(llm_request, "config") and llm_request.config:
            sys_inst = getattr(llm_request.config, "system_instruction", None)

        messages = self._map_messages(llm_request.contents, sys_inst)
        
        tools = None
        if hasattr(llm_request, "config") and llm_request.config:
            tools = getattr(llm_request.config, "tools", None)
        mapped_tools = self._map_tools(tools)

        openrouter_url = os.getenv("OPENROUTER_URL", "https://openrouter.ai").rstrip("/")
        api_key = os.getenv("OPENROUTER_API_KEY")

        if not api_key:
            # Fallback to Gemini / Google API key if OpenRouter key is not set and model is gemini
            api_key = os.getenv("GOOGLE_GENAI_API_KEY", "")

        model_name = self.model
        if model_name.startswith("gemini-"):
            model_name = model_name[7:]

        body: Dict[str, Any] = {
            "model": model_name,
            "messages": messages,
            "stream": stream,
        }

        if mapped_tools:
            body["tools"] = mapped_tools

        if hasattr(llm_request, "config") and llm_request.config:
            mime = getattr(llm_request.config, "response_mime_type", None)
            if mime == "application/json":
                body["response_format"] = {"type": "json_object"}

        headers = {
            "Content-Type": "application/json",
            "Authorization": f"Bearer {api_key}",
            "HTTP-Referer": "https://github.com/aaochoa/study-buddy",
            "X-Title": "Study Buddy AI Python",
        }

        return f"{openrouter_url}/api/v1/chat/completions", headers, body

    async def _handle_stream_response(self, response: httpx.Response) -> AsyncGenerator[LlmResponse, None]:
        accumulated_text = ""
        accumulated_tool_calls: Dict[int, Dict[str, Any]] = {}
        finish_reason = "STOP"

        async for line in response.aiter_lines():
            line = line.strip()
            if not line or not line.startswith("data: "):
                continue

            data_str = line[6:]
            if data_str == "[DONE]":
                break

            try:
                parsed = json.loads(data_str)
                choices = parsed.get("choices", [])
                if not choices:
                    continue
                
                choice = choices[0]
                if choice.get("finish_reason"):
                    finish_reason = "STOP" if choice["finish_reason"] == "stop" else choice["finish_reason"]

                delta = choice.get("delta", {})
                content_delta = delta.get("content")
                if content_delta:
                    accumulated_text += content_delta
                    part = types.Part.from_text(text=content_delta)
                    yield LlmResponse(
                        content=types.Content(role="model", parts=[part]),
                        partial=True
                    )

                tool_calls_delta = delta.get("tool_calls", [])
                for tc in tool_calls_delta:
                    idx = tc.get("index", 0)
                    if idx not in accumulated_tool_calls:
                        accumulated_tool_calls[idx] = {"arguments": ""}
                    acc = accumulated_tool_calls[idx]
                    if tc.get("id"):
                        acc["id"] = tc["id"]
                    fn = tc.get("function", {})
                    if fn.get("name"):
                        acc["name"] = fn["name"]
                    if fn.get("arguments"):
                        acc["arguments"] += fn["arguments"]

            except Exception:
                pass

        # Handle tool calls
        final_parts = []
        for idx in sorted(accumulated_tool_calls.keys()):
            acc = accumulated_tool_calls[idx]
            if acc.get("name"):
                try:
                    args = json.loads(acc["arguments"])
                except Exception:
                    args = {}
                fn_call = types.FunctionCall(name=acc["name"], args=args, id=acc.get("id"))
                final_parts.append(types.Part(function_call=fn_call))

        if not final_parts and accumulated_text:
            final_parts.append(types.Part.from_text(text=accumulated_text))

        if final_parts:
            yield LlmResponse(
                content=types.Content(role="model", parts=final_parts),
                partial=False
            )

    def _handle_single_response(self, data: dict) -> List[LlmResponse]:
        choices = data.get("choices", [])
        if not choices:
            raise RuntimeError(f"No choices returned from OpenRouter API: {data}")

        choice = choices[0]
        message = choice.get("message", {})
        parts = []

        if message.get("content"):
            parts.append(types.Part.from_text(text=message["content"]))

        for tc in message.get("tool_calls", []):
            if tc.get("type") == "function":
                fn = tc.get("function", {})
                try:
                    args = json.loads(fn.get("arguments", "{}"))
                except Exception:
                    args = {}
                fn_call = types.FunctionCall(name=fn.get("name", ""), args=args, id=tc.get("id"))
                parts.append(types.Part(function_call=fn_call))

        return [
            LlmResponse(
                content=types.Content(role="model", parts=parts),
                partial=False
            )
        ]

# Register OpenRouterLlm provider
LLMRegistry.register(OpenRouterLlm)
